### PR TITLE
Fix issue with stale cached data in LSP pull diagnostics

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Observed while dogfooding.  Removing this code has been smoke tested to confirm it fixes the issue.  ~~Currently investigating to make sure the removal of this code doesn't regress the original bug it was trying to fix.~~

Have confirmed that https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1618061 (the original bug you were fixing @mavasani ) still has the right behavior after this change.

### CODE REVIEW

File: src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs [hunk 1]
```diff
@@ -456,13 +449,6 @@ public async Task<bool> TryGetAsync(ArrayBuilder<DiagnosticData> list, Cancellat
                 {
                     var diagnostics = diagnosticsMap[analyzerWithState.Analyzer];
                     builder.AddRange(diagnostics.Where(ShouldInclude));
-
-                    // Save the computed diagnostics if caching is enabled and diagnostics were computed for the entire document.
-                    if (_cacheFullDocumentDiagnostics && !span.HasValue)
-                    {
-                        var data = new DocumentAnalysisData(version, _text.Lines.Count, diagnostics);
-                        analyzerWithState.State.Save(executor.AnalysisScope.Kind, data);
-                    }
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs:465]: here's why i think this is ok to remove.

1. this code doesn't run in normal VS scenarios (where everything is working fine).
2. it *only* runs in LSP, and LSP is the only case where things are buggy.  So we should just align both on the behavior that is non-buggy.

Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs:465]: If we remove this -  anything else calling `State.Save`?  It looks like some event on document open and [AnalyzeDocumentForKindAsync](https://sourceroslyn.io/Microsoft.CodeAnalysis.LanguageServer.Protocol/R/cb0a140fe44c3bde.html)

Its a bit hard to track down but is that only from the sln crawler?

Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs:465]: i'll see if i can clean this up more.  but i wouldn't mind the PR being minimal.

File: src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs [hunk 2]
```diff
@@ -456,13 +449,6 @@ public async Task<bool> TryGetAsync(ArrayBuilder<DiagnosticData> list, Cancellat
                 {
                     var diagnostics = diagnosticsMap[analyzerWithState.Analyzer];
                     builder.AddRange(diagnostics.Where(ShouldInclude));
-
-                    // Save the computed diagnostics if caching is enabled and diagnostics were computed for the entire document.
-                    if (_cacheFullDocumentDiagnostics && !span.HasValue)
```
Comment (mavasani) [inline src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs:461]: Note this was purely a performance improvement added for LSP pull diagnostics as path as all code paths computing full document diagnostics (tagger, squiggles, etc.) were going through this path and were recomputing the same set of document diagnostics.

Comment (mavasani) [reply src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs:461]: We avoided doing this caching for non-LSP path as diagnostic incremental analyzer executed from solution crawler has logic to push diagnostics to error list from background analysis path, and if we ended caching diagnostics from other path, such as tagger, error list wasn't getting properly refreshed. So, it felt easier to just do this `_cacheFullDocumentDiagnostics` logic here for LSP path.

File: src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs (hunk 1 above)
Comment (mavasani) [reply src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs:465]: What is the root cause of stale diagnostics here?

Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs:465]: no clue.  but we've confirmed on the LSP side that we're passing in the right document snapshot :)   so something *within* the diagnostics layer is caching and causing problems.  

As this is not a problem outside of LSP (where we are not doing the caching) and only a problem within LSP (where we are currently not doing it), the easiest thing was to just stop doing it :)

if you want to investigate the caching problem, i'm def ok with that.  but i'd like us to get into a good state first :)

File: src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs (hunk 2 above)
Comment (mavasani) [reply src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs:461]: > So if we were to add this back, i'd far prefer this caching just be in the LSP layer

That sounds reasonable to me.

> We need to start with correctness first, then work on optimizations.

I agree. Though I am still not able to follow the actual correctness issue here. We are caching full document diagnostics, just like we'd do in solution crawler path, and we don't have any known issues in the latter pertaining to caching full document diagnostics. One difference though is solution crawler executes callbacks into diagnostic incremental analyzer in a serial fashion, while the code path being removed in this PR can be executed concurrently from multiple clients.

> Note: this is the only issue currently blocking being able to enable pull diagnostics, so i'd like to make progress on this asap. Thanks!

Given this is just a performance aspect, I'd be fine to move ahead with a tracking issue filed to add this back. I can look next week and investigate this further.

Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs:461]: > Though I am still not able to follow the actual correctness issue here

Me neither. But it's definitely there :-)

**Comment by CyrusNajmabadi** [comment]
@mavasani let me know what the best way is to proceed here.

**Comment by mavasani** [comment]
> @mavasani let me know what the best way is to proceed here.

Oh sorry, I meant to suggest it is okay to merge this in and file a tracking issue to root cause the problem.

**Review by mavasani — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
> Oh sorry, I meant to suggest it is okay to merge this in and file a tracking issue to root cause the problem.

Will do.

**Pull request merged**